### PR TITLE
[SPARK-25747][SQL] remove ColumnarBatchScan.needsUnsafeRowConversion

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

`needsUnsafeRowConversion` is used in 2 places:
1. `ColumnarBatchScan.produceRows`
2. `FileSourceScanExec.doExecute`

When we hit `ColumnarBatchScan.produceRows`, it means whole stage codegen is on but the vectorized reader is off. The vectorized reader can be off for several reasons:
1. the file format doesn't have a vectorized reader(json, csv, etc.)
2. the vectorized reader config is off
3. the schema is not supported

Anyway when the vectorized reader is off, file format reader will always return unsafe rows, and other `ColumnarBatchScan` implementations also always return unsafe rows, so `ColumnarBatchScan.needsUnsafeRowConversion` is not needed.

When we hit `FileSourceScanExec.doExecute`, it means whole stage codegen is off. For this case, we need the `needsUnsafeRowConversion` to convert `ColumnarRow` to `UnsafeRow`, if the file format reader returns batch.

This PR removes `ColumnarBatchScan.needsUnsafeRowConversion`, and keep this flag only in `FileSourceScanExec`

## How was this patch tested?

existing tests

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @gatorsmile @hvanhovell

**Comment by cloud-fan** [comment]
retest this please

**Comment by kiszk** [comment]
Do we need to update the description? For example, `needsUnsafeRowConversion` exists in `DataSourceScanExec` now.

**Comment by cloud-fan** [comment]
which description is inaccurate?

**Comment by cloud-fan** [comment]
`DataSourceScanExec` does not have `needsUnsafeRowConversion`

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ColumnarBatchScan.scala [hunk 1]
```diff
@@ -164,12 +162,11 @@ private[sql] trait ColumnarBatchScan extends CodegenSupport {
     val outputVars = output.zipWithIndex.map { case (a, i) =>
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ColumnarBatchScan.scala]: set outputVars to null?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/ColumnarBatchScan.scala]: good catch!

**Comment by cloud-fan** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 2]
```diff
@@ -168,10 +168,11 @@ case class FileSourceScanExec(
...
   // so that this plan can be canonicalized on executor side too. See SPARK-23731.
-  override lazy val supportsBatch: Boolean = relation.fileFormat.supportBatch(
-    relation.sparkSession, StructType.fromAttributes(output))
+  override lazy val supportsBatch: Boolean = {
+    relation.fileFormat.supportBatch(relation.sparkSession, schema)
+  }
 
-  override lazy val needsUnsafeRowConversion: Boolean = {
+  private lazy val needsUnsafeRowConversion: Boolean = {
     if (relation.fileFormat.isInstanceOf[ParquetSource]) {
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: How about ORC? Why we only need to check the Parquet?

**Comment by gatorsmile** [comment]
LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: Our Parquet reader has one more feature than ORC: if vectorized reader is on but whole-stage-codegen is off, we can still read parquet with batches, and return `ColumnarRow`s. In ORC,  if whole-stage-codegen is off, we also turn off vectorized reader, so ORC never return `ColumnarRow`.

However, I just found that this parquet reader feature is gone. Let's send a new PR to fix it or completely remove it.

**Comment by kiszk** [comment]
retest this please

**Comment by kiszk** [comment]
You are right. Sorry, nvm
> DataSourceScanExec does not have needsUnsafeRowConversion

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Merged to master.

**Pull request closed without merging**